lib/cached_network_image.dart

```diff
@@ -86,81 +90,177 @@ class CacheManager {
 
-  SharedPreferences prefs;
-  Map<String, CacheObject> cacheData;
+  SharedPreferences _prefs;
+  Map<String, CacheObject> _cacheData;
+  DateTime lastCacheClean;
+
   static Object _lock = new Object();
 
   ///Shared preferences is used to keep track of the information about the files
   _init() async {
-    prefs = await SharedPreferences.getInstance();
+    _prefs = await SharedPreferences.getInstance();
 
     //get saved cache data from shared prefs
-    var jsonCacheString = prefs.getString("lib_cached_image_data");
-    cacheData = new Map();
+    var jsonCacheString = _prefs.getString("lib_cached_image_data");
+    _cacheData = new Map();
     if (jsonCacheString != null) {
       Map jsonCache = JSON.decode(jsonCacheString);
       jsonCache.forEach((key, data) {
-        cacheData[key] = new CacheObject.fromMap(data);
+        _cacheData[key] = new CacheObject.fromMap(key, data);
       });
     }
+
+    // Get data about when the last clean action has been performed
+    var cleanMillis = _prefs.getInt("lib_cached_image_data_last_clean");
+    if (cleanMillis != null) {
+      lastCacheClean = new DateTime.fromMillisecondsSinceEpoch(cleanMillis);
+    } else {
+      lastCacheClean = new DateTime.now();
+      _prefs.setInt("lib_cached_image_data_last_clean",
+          lastCacheClean.millisecondsSinceEpoch);
+    }
   }
 
+  bool _isStoringData = false;
+  bool _shouldStoreDataAgain = false;
+  Object _storeLock = new Object();
   ///Store all data to shared preferences
   _save() async {
+
+    if(!(await _canSave())){
+      return;
+    }
+
+    await _cleanCache();
+    await _saveDataInPrefs();
+  }
+
+  Future<bool> _canSave() async {
+    return await synchronized(_storeLock, (){
+      if(_isStoringData){
+        _shouldStoreDataAgain = true;
+        return false;
+      }
+      _isStoringData = true;
+      return true;
+    });
+  }
+
+  Future<bool> _shouldSaveAgain() async{
+    return await synchronized(_storeLock, (){
+      if(_shouldStoreDataAgain){
+        _shouldStoreDataAgain = false;
+        return true;
+      }
+      _isStoringData = false;
+      return false;
+    });
+  }
+
+  _saveDataInPrefs() async{
     Map json = new Map();
+
     await synchronized(_lock, () {
-      cacheData.forEach((key, cache) {
+      _cacheData.forEach((key, cache) {
         json[key] = cache._map;
       });
     });
-    prefs.setString("lib_cached_image_data", JSON.encode(json));
+    _prefs.setString("lib_cached_image_data", JSON.encode(json));
+
+    if(await _shouldSaveAgain()){
+      await _saveDataInPrefs();
+    }
+  }
+
+  _cleanCache({force: false}) async {
+    var sinceLastClean = new DateTime.now().difference(lastCacheClean);
+
+    if (force ||
+        sinceLastClean > inbetweenCleans ||
+        _cacheData.length > maxNrOfCacheObjects) {
+      await synchronized(_lock, () async {
+        var oldestDateAllowed = new DateTime.now().subtract(maxAgeCacheObject);
+
+        //Remove old objects
```

Comment:
can be function `_removeTooOldToKeepObjects` (i.e. comment will not be necessary)

Reply:
Again, I'll refactor it a bit.